services/explorer-indexer/src/modules/node/historicalSyncer.service.ts

```diff
@@ -0,0 +1,256 @@
+      };
+      verifiedTip = true;
+    }
+
+    while (!verifiedTip) {
+      const chainBlockAtSyncNumber = await this.rpcConnector.getBlockByNumber(
+        this.syncState.number
+      );
+      if (chainBlockAtSyncNumber?.hash !== this.syncState.hash) {
+        this.logger.log(
+          'Block hash at height',
+          this.syncState.number,
+          'has changed. Re-org detected, walking back.'
+        );
+        const lastBlock = await this.pbftService.getBlockByHash(
+          this.syncState.hash
+        );
+        // go back a step
+        this.syncState.number = Number(lastBlock.number) - 1;
+        this.syncState.hash = lastBlock.parent;
+      } else {
+        verifiedTip = true;
+      }
+    }
+
+    // sync blocks to tip
+    while (this.syncState.number < this.chainState.number) {
+      this.logger.log(`Getting block ${this.syncState.number + 1}...`);
+      const block: IPBFT = await this.rpcConnector.getBlockByNumber(
+        this.syncState.number + 1
+      );
+
+      this.logger.log(
+        `Getting ${block.transactions.length} transactions for block ${
+          this.syncState.number + 1
+        }...`
+      );
+      const blockTxs = [];
+      for (let i = 0; i < block.transactions.length; i++) {
+        this.logger.log(
+          `Getting transaction #${i + 1} out of #${
+            block.transactions.length
+          } - ${block.transactions[i]}...`
+        );
+        const tx = await this.rpcConnector.getTransactionByHash(
+          block.transactions[i]?.hash
+        );
+        blockTxs.push(tx);
+      }
+
+      block.transactions = blockTxs;
+
+      if (!block.hash) {
+        continue;
+      }
+
+      const started = new Date();
+      const txHashes: string[] = [];
+      let blockReward = 0;
+
+      const minBlock = Object.assign({}, block);
+
+      // @note : Right now there is no way to get the genesis transactions to set:
+      // Initial validators, initial balances for delegators and faucet.
+      // big TODO
+
+      const txReceipts = await this.getTransactionReceipts(
+        block.transactions.map((tx: ITransaction) => tx.hash),
+        20
+      );
+
+      block.transactions.forEach((tx: ITransaction, idx: number) => {
+        txHashes.push(tx.hash);
+        const receipt = txReceipts[idx];
+        blockReward =
+          blockReward + Number(receipt.gasUsed) * Number(tx.gasPrice);
+      });
+
+      minBlock.reward = blockReward;
+      minBlock.transactions = block.transactions;
+
+      this.logger.log(`Finalized block ${minBlock.hash}`);
+
+      await Promise.all(
+        block.transactions?.map((tx) => this.txService.safeSaveTransaction(tx))
+      );
+
+      await this.pbftService.safeSavePbft(block);
+
+      // update sync state
+      this.syncState.number = block.number;
+      this.syncState.genesis = this.chainState.genesis;
+      this.syncState.hash = block.hash;
+
+      const completed = new Date();
+      console.log(
```

Comment:
Shouldn't we add logger here as well instead of console.log?

Reply:
Yes, it needs to be siwtched to logger everywhere, I just need console so that it prints during debug